### PR TITLE
Implement selectors

### PR DESCRIPTION
**Pull request by kenzierocks**
See https://github.com/SpongePowered/Sponge/pull/243.

### CODE REVIEW

**Comment by zml2008** [comment]
This needs to be updated to set the appropriate fields in the GameRegistry, because currently none of this stuff is hooked up.

**Comment by kenzierocks** [comment]
Looks like I lost that in the copy. Porting ASAP.

**Comment by kenzierocks** [comment]
Alright, I think that should be good. If you merge the API and update the ref on master, I'll rebase this.

File: src/main/java/org/spongepowered/common/text/selector/SpongeSelectorFactory.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        if (!selector.startsWith("@")) {
+            throw new IllegalArgumentException("Invalid selector " + selector);
+        }
+        // If multi-character types are possible, this handles it
+        int argListIndex = selector.indexOf('[');
+        if (argListIndex < 0) {
+            argListIndex = selector.length();
+        }
+        String typeStr = selector.substring(1, argListIndex);
+        Optional<SelectorType> type = Sponge.getGame().getRegistry().getType(SelectorType.class, typeStr);
```
Comment (zml2008) [inline src/main/java/org/spongepowered/common/text/selector/SpongeSelectorFactory.java]: This part won't work with the keys you registered selectors under though?

File: src/main/resources/common_at.cfg [hunk 2]
```diff
@@ -109,3 +109,5 @@ public net.minecraft.entity.EntityLivingBase field_70755_b # entityLivingToAttac
 public-f net.minecraft.entity.EntityLivingBase field_94063_bt # _combatTracker
 
 public net.minecraft.util.FoodStats field_75126_c # foodExhaustionLevel
+
```
Comment (zml2008) [inline src/main/resources/common_at.cfg]: Also you have to update the AT. Please actually test this...

File: src/main/java/org/spongepowered/common/text/selector/SpongeSelectorFactory.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T, V> Set<Argument<T>> createArguments(ArgumentHolder<? extends ArgumentType<T>> type, V value) {
+        Set<Argument<T>> set = Sets.newLinkedHashSet();
+        if (type instanceof SpongeArgumentHolder.SpongeVector3) {
+            Set<Function<V, T>> extractors = ((SpongeArgumentHolder.SpongeVector3<V, T>) (Object) type).extractFunctions();
+            Set<? extends ArgumentType<T>> types = type.getTypes();
+            Iterator<Function<V, T>> extIter = extractors.iterator();
+            Iterator<? extends ArgumentType<T>> typeIter = types.iterator();
+            for (; extIter.hasNext() && typeIter.hasNext();) {
```
Comment (Mumfrey) [inline src/main/java/org/spongepowered/common/text/selector/SpongeSelectorFactory.java]: This is normally called a `while` loop. Why the bizarre use of `for` ?

Comment (kenzierocks) [inline src/main/java/org/spongepowered/common/text/selector/SpongeSelectorFactory.java]: I think I used to initialize the iterators in it. I'll patch it.

**Comment by Aaron1011** [comment]
@kenzierocks: I'd recommend commiting on a branch other then `master`. I'll save you a lot of headache when pulling in other changes.

**Comment by kenzierocks** [comment]
I normally do, this was done in a rush.

**Comment by zml2008** [comment]
Could you try to get this fixed up @kenzierocks?

**Comment by kenzierocks** [comment]
Nuking this in favor of new branch,

**Pull request closed without merging**